File: runtime/ts/src/callable_identity.test.ts
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { test } from 'node:test';
import { callableIdentity } from './callable_identity.ts';
import { canonicalDeclaration, hashDeclaration, typeDeclaration, withDeclaration } from './declaration_identity.ts';
import { createValidator } from './validate.ts';
import type { Slots, TypeBinding, TypeExpression, Validator, WireFamily } from './validate.ts';

interface Source {
  name: string;
  family: string;
  wireExpression: TypeExpression;
  families: Record<string, { wire: string; declaration: string; digest: string; imports: string[] }>;
}
interface Slot {
  source?: string;
  type?: TypeExpression;
  slots?: Record<string, Slot>;
}
interface Case {
  id: string;
  source: string;
  slots?: Record<string, Slot>;
  path?: string;
  digest?: string;
  error?: string;
}
const sources = new Map<string, Source>(
  (
    JSON.parse(
      readFileSync(new URL('../../../conformance/tables/declaration-digests.json', import.meta.url), 'utf8'),
    ) as { cases: Source[] }
  ).cases.map((row) => [row.name, row]),
);
const table = JSON.parse(
  readFileSync(new URL('../../../conformance/tables/callable-identities.json', import.meta.url), 'utf8'),
) as { cases: Case[]; families: Record<string, { wire: WireFamily; declaration: unknown; imports?: string[] }> };

function fixture(source: string, type?: TypeExpression, slots: Record<string, Slot> = {}): TypeBinding {
  const row = sources.get(source);
  assert.ok(row, source);
  const validators: Record<string, Validator> = {},
    imports: Record<string, Record<string, Validator>> = {};
  for (const [name, family] of Object.entries(row.families)) {
    imports[name] = {};
    validators[name] = withDeclaration(
      createValidator(JSON.parse(family.wire) as WireFamily, family.digest, imports[name]),
      family.declaration,
    );
  }
  for (const [name, family] of Object.entries(row.families))
    for (const imported of family.imports) imports[name]![imported] = validators[imported]!;
  const bindings: Record<string, TypeBinding> = {};
  for (const [name, slot] of Object.entries(slots))
    bindings[name] = fixture(slot.source ?? source, slot.type, slot.slots);
  return { validate: validators[row.family]!, type: type ?? row.wireExpression, slots: bindings };
}

for (const row of table.cases)
  test(row.id, () => {
    const binding = fixture(row.source, undefined, row.slots);
    if (row.error) {
      assert.throws(
        () => callableIdentity(binding),
        (error) => error instanceof Error && error.message.includes(row.error!),
      );
      return;
    }
    const identity = callableIdentity(binding);
    assert.equal(identity.path, row.path);
    if (row.digest) assert.equal(identity.digest, row.digest);
    assert.equal(identity.digest, hashDeclaration(typeDeclaration(binding)));
    const reference: Record<string, unknown> = { binding: 'opaque', contract: identity.path, digest: identity.digest };
    binding.validate(binding.type, reference, '$', binding.slots);
    delete reference.digest;
    binding.validate(binding.type, reference, '$', binding.slots);
    reference.contract = 'same/Function';
    assert.throws(() => binding.validate(binding.type, reference, '$', binding.slots));
  });

function customFixtures(): Record<string, Validator> {
  const validators: Record<string, Validator> = {},
    imports: Record<string, Record<string, Validator>> = {};
  for (const [name, family] of Object.entries(table.families)) {
    imports[name] = {};
    const document = canonicalDeclaration(family.declaration);
    validators[name] = withDeclaration(
      createValidator(family.wire, hashDeclaration(document), imports[name]),
      document,
    );
  }
  for (const [name, family] of Object.entries(table.families))
    for (const imported of family.imports ?? []) imports[name]![imported] = validators[imported]!;
  return validators;
}

test('GEN-ID-CAPTURED-SCOPE preserves imported application and alias provenance', () => {
  const validators = customFixtures(),
    captured = validators.captured!,
    consumer = validators.consumer!;
  const slots: Slots = { T: { validate: captured, type: 'string' } };
  const binding: TypeBinding = { validate: captured, type: 'Callback', slots };
  const identity = callableIdentity(binding);
  assert.equal(identity.path, 'captured/Callback<string>');
  for (const type of ['Closed', { apply: 'captured.Callback', with: { T: 'string' } }] as TypeExpression[]) {
    assert.deepEqual(callableIdentity({ validate: consumer, type }), identity);
    consumer(type, { binding: 'opaque', contract: identity.path, digest: identity.digest });
  }
  assert.throws(() => callableIdentity({ validate: captured, type: 'Callback' }), /binding T/);
  assert.throws(() => captured('Callback', { binding: 'opaque', contract: 'captured/Callback' }), /binding T/);
  const plain = callableIdentity({ validate: consumer, type: 'Alias' });
  assert.equal(plain.path, 'captured/Plain');
  assert.equal(plain.digest, hashDeclaration(canonicalDeclaration(table.families.captured!.declaration)));
  assert.notEqual(plain.digest, hashDeclaration(canonicalDeclaration(table.families.consumer!.declaration)));
  const other = callableIdentity({
    validate: captured,
    type: 'Callback',
    slots: { T: { validate: captured, type: 'integer' } },
  });
  assert.notEqual(other.path, identity.path);
  assert.notEqual(other.digest, identity.digest);
  assert.throws(() => captured('Callback', { binding: 'opaque', contract: other.path }, '$', slots));
  const entity = callableIdentity({
    validate: captured,
    type: 'Callback',
    slots: { T: { validate: captured, type: { ref: 'Entity' } } },
  });
  assert.equal(entity.path, 'captured/Callback<&captured/Entity>');
  const drawn: TypeBinding = {
    validate: validators.drawn!,
    type: 'Callback',
    slots: { S: { name: 'captured', validate: captured, slots } },
  };
  const drawIdentity = callableIdentity(drawn);
  assert.equal(drawIdentity.path, 'drawn/Callback<captured<string>>');
  drawn.validate(
    drawn.type,
    { binding: 'opaque', contract: drawIdentity.path, digest: drawIdentity.digest },
    '$',
    drawn.slots,
  );
});

test('GEN-ID-REVISION-REFUSAL preserves the optional digest boundary', () => {
  const first = fixture('application argument declaration baseline'),
    second = fixture('application reachable argument changed');
  const identity = callableIdentity(first);
  assert.throws(
    () => second.validate(second.type, { binding: 'opaque', contract: identity.path, digest: identity.digest }),
    { code: 'contract_mismatch' },
  );
  second.validate(second.type, { binding: 'opaque', contract: identity.path });
  assert.throws(
    () =>
      callableIdentity({
        validate: createValidator({ types: { Plain: { kind: 'callable', contract: 'x/Plain' } } }, ''),
        type: 'Plain',
      }),
    /provenance/,
  );
});

File: runtime/go/callable_identity_test.go
package runtime

import (
	"encoding/json"
	"errors"
	"os"
	"strings"
	"testing"
)

func customCallableFixtures(t *testing.T) map[string]*Schema {
	t.Helper()
	data, err := os.ReadFile("../../conformance/tables/callable-identities.json")
	if err != nil {
		t.Fatal(err)
	}
	var table struct {
		Families map[string]struct {
			Wire        any
			Declaration any
			Imports     []string
		}
	}
	if err := json.Unmarshal(data, &table); err != nil {
		t.Fatal(err)
	}
	result := map[string]*Schema{}
	for name, fixture := range table.Families {
		wire, _ := json.Marshal(fixture.Wire)
		declaration, _ := json.Marshal(fixture.Declaration)
		result[name] = MustSchema(string(wire), declarationHash(string(declaration)), nil).MustWithDeclaration(string(declaration))
	}
	for name, fixture := range table.Families {
		for _, imported := range fixture.Imports {
			result[name].imported[imported] = result[imported]
		}
	}
	return result
}

func TestCallableIdentityCapturedScopesAndAliases(t *testing.T) {
	schemas := customCallableFixtures(t)
	captured, consumer := schemas["captured"], schemas["consumer"]
	integer := TypeBinding{Schema: captured, Type: "integer"}
	text := TypeBinding{Schema: captured, Type: "string"}
	bound := TypeBinding{Schema: captured.Bind(map[string]any{"T": text}, nil), Type: "Callback"}
	identity, err := CallableIdentity(bound)
	if err != nil || identity.Path != "captured/Callback<string>" {
		t.Fatalf("capture: %v %v", identity, err)
	}
	for _, binding := range []TypeBinding{{Schema: consumer, Type: "Closed"}, {Schema: consumer, Type: MustTypeExpression(`{"apply":"captured.Callback","with":{"T":"string"}}`)}} {
		got, err := CallableIdentity(binding)
		if err != nil || got != identity {
			t.Fatalf("imported capture/alias differs: %v %v", got, err)
		}
		if err := validateCallableReference(binding, map[string]any{"binding": "opaque", "contract": identity.Path, "digest": identity.Digest}); err != nil {
			t.Fatal(err)
		}
	}
	if _, err := CallableIdentity(TypeBinding{Schema: captured, Type: "Callback"}); err == nil {
		t.Fatal("unbound capture accepted")
	}
	if err := validateCallableReference(TypeBinding{Schema: captured, Type: "Callback"}, map[string]any{"binding": "opaque", "contract": "captured/Callback"}); err == nil {
		t.Fatal("unbound capture accepted by validator")
	}
	plain, err := CallableIdentity(TypeBinding{Schema: consumer, Type: "Alias"})
	if err != nil || plain.Path != "captured/Plain" || plain.Digest != captured.digest || plain.Digest == consumer.digest {
		t.Fatalf("nongeneric alias: %v %v", plain, err)
	}
	other, err := CallableIdentity(TypeBinding{Schema: captured.Bind(map[string]any{"T": integer}, nil), Type: "Callback"})
	if err != nil || other.Path == identity.Path || other.Digest == identity.Digest {
		t.Fatalf("capture collapsed: %v %v", other, err)
	}
	if err := validateCallableReference(bound, map[string]any{"binding": "opaque", "contract": other.Path}); err == nil {
		t.Fatal("wrong capture accepted without digest")
	}
	entityKey := TypeBinding{Schema: captured, Type: MustTypeExpression(`{"ref":"Entity"}`)}
	entity, err := CallableIdentity(TypeBinding{Schema: captured.Bind(map[string]any{"T": entityKey}, nil), Type: "Callback"})
	if err != nil || entity.Path != "captured/Callback<&captured/Entity>" {
		t.Fatalf("entity-key argument: %v %v", entity, err)
	}
	drawn := TypeBinding{Schema: schemas["drawn"].Bind(nil, map[string]*Schema{"S": bound.Schema}), Type: "Callback"}
	drawIdentity, err := CallableIdentity(drawn)
	if err != nil || drawIdentity.Path != "drawn/Callback<captured<string>>" {
		t.Fatalf("bound family capture: %v %v", drawIdentity, err)
	}
	if err := validateCallableReference(drawn, map[string]any{"binding": "opaque", "contract": drawIdentity.Path, "digest": drawIdentity.Digest}); err != nil {
		t.Fatal(err)
	}
}

func TestCallableIdentityRevisionRefusalBeforeInterpretation(t *testing.T) {
	rows := identityFixtures(t)
	first := callableFixtureBinding(t, rows, "application argument declaration baseline", nil, nil)
	second := callableFixtureBinding(t, rows, "application reachable argument changed", nil, nil)
	identity, err := CallableIdentity(first)
	if err != nil {
		t.Fatal(err)
	}
	err = validateCallableReference(second, map[string]any{"binding": "opaque", "contract": identity.Path, "digest": identity.Digest})
	var public *PublicError
	if !errors.As(err, &public) || public.Code != "contract_mismatch" {
		t.Fatalf("revision refusal: %v", err)
	}
	if err := validateCallableReference(second, map[string]any{"binding": "opaque", "contract": identity.Path}); err != nil {
		t.Fatalf("unspecified revision: %v", err)
	}
	if _, err := CallableIdentity(TypeBinding{Schema: MustSchema(`{"types":{"Plain":{"kind":"callable","contract":"x/Plain"}}}`, "", nil), Type: "Plain"}); err == nil {
		t.Fatal("public identity accepted absent provenance")
	}
}

type callableIdentitySlot struct {
	Source string                          `json:"source"`
	Type   json.RawMessage                 `json:"type"`
	Slots  map[string]callableIdentitySlot `json:"slots"`
}

type callableIdentityCase struct {
	ID, Source, Path, Digest, Error string
	Slots                           map[string]callableIdentitySlot
}

func validateCallableReference(binding TypeBinding, reference map[string]any) error {
	raw, err := json.Marshal(reference)
	if err != nil {
		return err
	}
	return binding.Schema.ValidateExpressionRaw(binding.Type, raw)
}

func callableFixtureBinding(t *testing.T, rows map[string]identityFixture, source string, typ json.RawMessage, slots map[string]callableIdentitySlot) TypeBinding {
	t.Helper()
	row := rows[source]
	if row.Name == "" {
		t.Fatalf("missing declaration fixture %s", source)
	}
	schemas := map[string]*Schema{}
	for name, family := range row.Families {
		schemas[name] = MustSchema(family.Wire, family.Digest, nil).MustWithDeclaration(family.Declaration)
	}
	for name, family := range row.Families {
		for _, imported := range family.Imports {
			schemas[name].imported[imported] = schemas[imported]
		}
	}
	bound := map[string]any{}
	for name, slot := range slots {
		origin := slot.Source
		if origin == "" {
			origin = source
		}
		bound[name] = callableFixtureBinding(t, rows, origin, slot.Type, slot.Slots)
	}
	if len(typ) == 0 {
		typ = row.WireExpression
	}
	return TypeBinding{Schema: schemas[row.Family].Bind(bound, nil), Type: MustTypeExpression(string(typ))}
}

func TestCallableIdentitySharedPaths(t *testing.T) {
	data, err := os.ReadFile("../../conformance/tables/callable-identities.json")
	if err != nil {
		t.Fatal(err)
	}
	var table struct{ Cases []callableIdentityCase }
	if err := json.Unmarshal(data, &table); err != nil {
		t.Fatal(err)
	}
	rows := identityFixtures(t)
	for _, row := range table.Cases {
		t.Run(row.ID, func(t *testing.T) {
			binding := callableFixtureBinding(t, rows, row.Source, nil, row.Slots)
			got, err := CallableIdentity(binding)
			if row.Error != "" {
				if err == nil || !strings.Contains(err.Error(), row.Error) {
					t.Fatalf("got %v, want %s", err, row.Error)
				}
				return
			}
			if err != nil {
				t.Fatal(err)
			}
			if got.Path != row.Path {
				t.Fatalf("path\ngot  %s\nwant %s", got.Path, row.Path)
			}
			if row.Digest != "" && got.Digest != row.Digest {
				t.Fatalf("digest got %s, want %s", got.Digest, row.Digest)
			}
			declaration, err := binding.Declaration()
			if err != nil || got.Digest != declarationHash(declaration) {
				t.Fatalf("applied digest does not identify selected graph: %v", err)
			}
			reference := map[string]any{"binding": "opaque", "contract": got.Path, "digest": got.Digest}
			if err := validateCallableReference(binding, reference); err != nil {
				t.Fatalf("own reference: %v", err)
			}
			delete(reference, "digest")
			if err := validateCallableReference(binding, reference); err != nil {
				t.Fatalf("absent revision: %v", err)
			}
			reference["contract"] = "same/Function"
			if err := validateCallableReference(binding, reference); err == nil {
				t.Fatal("template contract bypassed applied nominal check without digest")
			}
		})
	}
}
